csharp/autobuilder/Semmle.Autobuild.CSharp/DotNetRule.cs

```diff
@@ -147,11 +140,59 @@ private static BuildScript DownloadDotNet(IAutobuilder<AutobuildOptionsShared> b
                     continue;
                 }
 
-                installScript &= DownloadDotNetVersion(builder, installDir, version);
-                validGlobalJson = true;
+                return version;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns a script for downloading relevant versions of the
+        /// .NET Core SDK. The SDK(s) will be installed at <code>installDir</code>
+        /// (provided that the script succeeds).
+        /// </summary>
+        private static BuildScript DownloadDotNet(IAutobuilder<AutobuildOptionsShared> builder, string installDir)
+        {
+            if (!string.IsNullOrEmpty(builder.Options.DotNetVersion))
+                // Specific version supplied in configuration: always use that
+                return DownloadDotNetVersion(builder, installDir, builder.Options.DotNetVersion);
+
+            // Download versions mentioned in `global.json` files
+            // See https://docs.microsoft.com/en-us/dotnet/core/tools/global-json
+            var installScript = BuildScript.Success;
+            var foundFrameworkVersion = false;
+            var globalJsonSdkVersion = SdkVersionFromGlobalJson(builder);
+
+            if (!string.IsNullOrEmpty(globalJsonSdkVersion))
+            {
+                installScript &= DownloadDotNetVersion(builder, installDir, globalJsonSdkVersion);
+                foundFrameworkVersion = true;
+            }
+            else
+            {
+                // If there is no `global.json` to retrieve the .NET version from, we find all the
+                // TargetFramework values in the project files and install the latest versions
+                // from the corresponding channels
+                var targetFrameworkVersions = builder.ProjectsOrSolutionsToBuild
+                    .SelectMany(p => Enumerators.Singleton(p).Concat(p.IncludedProjects))
+                    .OfType<Project<CSharpAutobuildOptions>>()
+                    .Select(p => p.TargetFramework)
+                    .OfType<string>()
+                    .Where(v => v.StartsWith("net"));
```

Comment:
I think we could be more specific about the versions. There's no point in issuing a web request for things that will obviously fail, such as `netstandard.*`, `net48` and below, and maybe `netcoreapp3.1` and below also doesn't work.

Reply:
`netstandard` and `netcoreapp` should already get filtered out by the code further down (see the comment there). There was a unit test which initially failed for me (because it specified `netcoreapp2.1` or something), but now passes since we no longer attempt to install those SDKs.

I can also filter out older `net` versions.

Reply:
Maybe it's as simple as filtering out anything that has a major version number below 5.